File: lib/api/assistant.ts
import { AuthApiError } from "@/lib/api/auth";
import { apiUrl } from "@/lib/api/base";
import type { Citation } from "@/types/dashboard";

// ── Types ─────────────────────────────────────────────────────────────────────

export interface AgentStep {
  stepIndex: number;
  stepType: "planner" | "retriever" | "scorer" | "generator" | "verifier" | "decider" | string;
  inputJson?: Record<string, unknown> | null;
  outputJson?: Record<string, unknown> | null;
  elapsedMs?: number | null;
}

interface AssistantAskPayload {
  chatSessionId: string;
  traceId: string;
  answer: string;
  confidence: number;
  citations: Citation[];
  agentSteps: AgentStep[];
}

interface AssistantSessionSummaryResponse {
  id: string;
  title: string;
  createdAt: string;
  updatedAt: string;
  documentCount: number;
  topicHints: string[];
}

// ── Helpers ──────────────────────────────────────────────────────────────────

async function parseResponse<T>(response: Response): Promise<T> {
  const contentType = response.headers.get("content-type") ?? "";
  const payload = contentType.includes("application/json")
    ? await response.json()
    : null;

  if (!response.ok) {
    const detail = payload?.detail;
    const message =
      typeof detail === "string"
        ? detail
        : detail?.message || payload?.message || "Request failed";
    throw new AuthApiError(message, response.status, detail);
  }

  return payload as T;
}

// ── Endpoints ────────────────────────────────────────────────────────────────

export async function askAssistant(params: {
  question: string;
  chatSessionId?: string | null;
  documentIds?: string[] | null;
  topK?: number;
  mode?: "rag" | "medical_chat";
}): Promise<{
  chatSessionId: string;
  traceId: string;
  message: string;
  confidence: number;
  citations: Citation[];
  agentSteps: AgentStep[];
}> {
  const response = await fetch(apiUrl("/assistant/ask"), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify({
      question: params.question,
      chatSessionId: params.chatSessionId ?? undefined,
      documentIds: params.documentIds ?? undefined,
      topK: params.topK ?? 6,
      mode: params.mode ?? "rag",
    }),
  });

  const payload = await parseResponse<AssistantAskPayload>(response);

  return {
    chatSessionId: payload.chatSessionId,
    traceId: payload.traceId,
    message: payload.answer,
    confidence: payload.confidence,
    citations: payload.citations,
    agentSteps: payload.agentSteps ?? [],
  };
}

export interface AssistantSessionSummary {
  id: string;
  title: string;
  createdAt: string;
  updatedAt: string;
  documentCount: number;
  topicHints: string[];
}

export interface AssistantSessionMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  timestamp: string;
  citations?: Citation[];
  confidence?: number | null;
  traceId?: string | null;
  agentSteps?: AgentStep[];
}

export interface AssistantSessionDetail {
  id: string;
  title: string;
  createdAt: string;
  updatedAt: string;
  messages: AssistantSessionMessage[];
}

export async function listAssistantSessions(): Promise<AssistantSessionSummary[]> {
  const response = await fetch(apiUrl("/assistant/sessions"), {
    credentials: "include",
    cache: "no-store",
  });
  return parseResponse<AssistantSessionSummaryResponse[]>(response);
}

export async function getAssistantSession(
  sessionId: string
): Promise<AssistantSessionDetail> {
  const response = await fetch(apiUrl(`/assistant/sessions/${sessionId}`), {
    credentials: "include",
    cache: "no-store",
  });
  return parseResponse<AssistantSessionDetail>(response);
}

File: backend/app/schemas/assistant.py
from __future__ import annotations

from typing import Annotated, Any

from pydantic import BaseModel, ConfigDict, Field


class AssistantMode(str):
    RAG = "rag"
    MEDICAL_CHAT = "medical_chat"


class AssistantCitation(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    document_name: str = Field(alias="documentName")
    page: int
    chapter: str
    snippet: str


class AssistantAskRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    question: str = Field(min_length=2, max_length=4000)
    chat_session_id: Annotated[str | None, Field(alias="chatSessionId")] = None
    top_k: Annotated[int, Field(ge=1, le=20, alias="topK")] = 10
    document_ids: Annotated[list[str] | None, Field(alias="documentIds")] = None
    mode: str = Field(default=AssistantMode.RAG, description="rag | medical_chat")


class AgentStepOut(BaseModel):
    """One agentic reasoning step returned in the /ask response."""
    model_config = ConfigDict(populate_by_name=True)

    step_index: int = Field(alias="stepIndex")
    step_type: str = Field(alias="stepType")
    input_json: dict[str, Any] | None = Field(default=None, alias="inputJson")
    output_json: dict[str, Any] | None = Field(default=None, alias="outputJson")
    elapsed_ms: int | None = Field(default=None, alias="elapsedMs")


class AssistantAskResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    chat_session_id: str = Field(alias="chatSessionId")
    trace_id: str = Field(alias="traceId")
    answer: str
    confidence: int
    citations: list[AssistantCitation]
    agent_steps: list[AgentStepOut] = Field(default_factory=list, alias="agentSteps")


class AssistantSessionSummary(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")
    document_count: int = Field(alias="documentCount")
    topic_hints: list[str] = Field(default_factory=list, alias="topicHints")


class AssistantSessionMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    role: str
    content: str
    timestamp: str
    citations: list[AssistantCitation] = Field(default_factory=list)
    confidence: int | None = None
    trace_id: str | None = Field(default=None, alias="traceId")
    agent_steps: list[AgentStepOut] = Field(default_factory=list, alias="agentSteps")


class AssistantSessionDetail(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")
    messages: list[AssistantSessionMessage]
